kubernetes/cache/cache.go

```diff
@@ -163,90 +207,126 @@ func (c *kialiCacheImpl) isCached(namespace string) bool {
+		c.stopNSChans[namespace] = make(chan struct{})
 	}
 
-	go func(stopCh <-chan struct{}) {
-		for _, informer := range c.nsCache[namespace] {
-			go informer.Run(stopCh)
-		}
-		<-stopCh
-		log.Infof("Kiali cache for [namespace: %s] stopped", namespace)
-	}(c.stopChan[namespace])
+	kubeInformerFactory.Start(c.stopNSChans[namespace])
+	istioInformerFactory.Start(c.stopNSChans[namespace])
 
 	log.Infof("Waiting for Kiali cache for [namespace: %s] to sync", namespace)
-	isSynced := func() bool {
-		hasSynced := true
-		for _, informer := range c.nsCache[namespace] {
-			hasSynced = hasSynced && informer.HasSynced()
+
+	for _, synced := range kubeInformerFactory.WaitForCacheSync(c.stopNSChans[namespace]) {
+		if !synced {
+			log.Error("[Kiali Cache] failed to sync")
+			return false
 		}
-		return hasSynced
 	}
-	if synced := cache.WaitForCacheSync(c.stopChan[namespace], isSynced); !synced {
-		c.stopChan[namespace] <- struct{}{}
-		log.Errorf("Kiali cache for [namespace: %s] sync failure", namespace)
-		return false
+	for _, synced := range istioInformerFactory.WaitForCacheSync(c.stopNSChans[namespace]) {
+		if !synced {
+			log.Error("[Kiali Cache] failed to sync")
+			return false
+		}
 	}
+
 	log.Infof("Kiali cache for [namespace: %s] started", namespace)
 
 	return true
 }
 
+func (c *kialiCacheImpl) createClusterScopedCache() bool {
+	kubeInformerFactory := c.createKubernetesInformers("")
+	istioInformerFactory := c.createIstioInformers("")
+
+	c.stopClusterScopedChan = make(chan struct{})
+	kubeInformerFactory.Start(c.stopClusterScopedChan)
+	istioInformerFactory.Start(c.stopClusterScopedChan)
+
+	log.Info("Waiting for Kiali cache for to sync")
+	for _, synced := range kubeInformerFactory.WaitForCacheSync(c.stopClusterScopedChan) {
+		if !synced {
+			log.Error("[Kiali Cache] failed to sync")
+			return false
+		}
+	}
+	for _, synced := range istioInformerFactory.WaitForCacheSync(c.stopClusterScopedChan) {
+		if !synced {
+			log.Error("[Kiali Cache] failed to sync")
+			return false
+		}
+	}
+
+	log.Info("[Kiali Cache] started")
+
+	return true
+}
+
 // CheckNamespace will
 // - Validate if a namespace is included in the cache
 // - Create and initialize a cache
-// - Validate if a cache is synced
 func (c *kialiCacheImpl) CheckNamespace(namespace string) bool {
-	if !c.isCached(namespace) {
+	if c.clusterScoped {
+		return true
+	} else if !c.isCached(namespace) {
 		return false
 	}
 
 	c.cacheLock.RLock()
-	_, isNsCached := c.nsCache[namespace]
+	_, isNsCached := c.nsCacheLister[namespace]
 	c.cacheLock.RUnlock()
 
 	if !isNsCached {
 		defer c.cacheLock.Unlock()
 		c.cacheLock.Lock()
-		return c.createCache(namespace)
+		return c.createNSCache(namespace)
 	}
-	return c.isKubernetesSynced(namespace) && c.isIstioSynced(namespace)
+
+	return true
 }
 
 // RefreshNamespace will delete the specific namespace's cache and create a new one.
+// If a cluster-scoped namespace is used, it will recreate the cluster-scoped cache.
```

Comment:
This seems strange, would it not lead to refreshing a cluster-scoped cache too often, in some loop for namespaces?  I think I'd rather see a dedicated function for each, and if the namespace version is called when cluster-scoped cache is enabled, it should be an error (and vice-versa).

Reply:
Yes the cluster scoped cache could get refreshed in a for loop but the refresh only happens when we do a write operation on some resource: https://github.com/kiali/kiali/blob/master/business/services.go#L595. Unless Kiali is updating resources across many namespaces at once, many different refreshes at once are not very likely to happen. There may be other places but the main spot that Kiali performs write operations is through the wizards.

My guess is that the primary purpose of the refresh is that there's no way, that I know of, to fast forward the cache to a particular ResourceVersion. You could update the object then do another read from the cache and not see the updated object. Recreating the cache entirely force a fast forward in a way to the latest ResourceVersion.

I'll look into adding separate functions for the refresh but it'd be nice for clients to not have to know that the cache is/isn't cluster scoped.

Reply:
If you feel like the direct callers of this function should not care about the cache-type, and that there is no real concern around the "unnecessary looping" comment, I'm fine with things as-is.  My comment was mainly around catching issues in the calling code where the cache-type was being mishandled.  I might just suggest, for clarity, a change to the function name, but totally up to you:

```
// RefreshCache will recreate the necessary cache. If the cache is cluster-scoped the "namespace" argument 
// is ignored, otherwise only the namespace-specific cache is updated.
func (c *kialiCacheImpl) RefreshCache(namespace string)
```

Reply:
Renaming sounds good. Direct callers would still need to call refresh in all the same places (writes to kube objects) regardless of whether the cache is namespace or cluster scoped so it would only change where the check for cluster vs. namespace scope is happening. Cluster scoped refreshes could be slower especially if there's a lot of kube objects to list. It's a problem for namespace scoped as well but might be worse with cluster scope since there's more objects. As far as I can tell writes are fairly infrequent operations though as they happen only when a user uses a wizard or adds injection etc. This is something that needs to be tested out with this change.